CRM Automations UI: Add workflow saving and reloading

robertf4 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/zero-bs-crm/issues/3203

## Proposed changes:
<!--- Explain what functional changes your PR includes -->

This PR implements the saving and reloading of workflows.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
NtYgueI8oAgiWGHTbHpnjh-fi-1_29

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Be sure you have test workflows on your site. If not then see this PR for instructions on making some: https://github.com/Automattic/jetpack/pull/33416.

1. Visit `/wp-admin/admin.php?page=jpcrm-automations#/automations`.
2. Toggle the active toggles on some of your workflows. Hard refresh the page to verify that the state is saved.
3. Select some workflows and use the bulk action to change their active state. Hard refresh the page to verify that the state is saved.
4. Click "Edit" for a workflow and change some attributes, then press "Save".  Hard refresh the page to verify that the state is saved.
5. Click "Edit" for a workflow, change some attributes, then press "Cancel". Now press "Edit" again and verify that your changes are gone. Hard refresh the page and click "Edit" again to verify that your changes were not saved.

robertf4 committed:
WIP

robertf4 committed:
changelog

kallehauge committed:
[CRM] Define attributes for most contact and invoice actions

robertf4 committed:
Remove findStep and add getServerPreparedWorkflow

robertf4 committed:
Use value in select inputs

robertf4 committed:
Update query functions

robertf4 committed:
Add onSave workflow dependency

robertf4 committed:
Implement edit modal canceling

robertf4 committed:
Save active state when toggle changed

robertf4 committed:
Implement bulk activate/deactivate loading and saving

robertf4 commented on projects/plugins/crm/src/js/components/automations-admin/components/attribute-config/index.tsx:
```diff
@@ -61,6 +61,7 @@ const getEditValue = (
 							  } ) )
 							: []
 					}
+					value={ value }
```
This just fixes a bug where on load the SelectControl would display the first option instead of the option selected in the store.

robertf4 commented on projects/plugins/crm/src/js/components/automations-admin/components/bulk-workflow-actions/index.tsx:
```diff
@@ -1,26 +1,61 @@
 import { DropdownMenu, MenuItem } from '@wordpress/components';
-import { dispatch } from '@wordpress/data';
+import { useSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
```
There were a few non-internationalized strings in here, so I went ahead and corrected that.

robertf4 commented on projects/plugins/crm/src/rest-api/v4/class-rest-automation-workflows-controller.php:
```diff
@@ -152,9 +152,22 @@ public function register_routes() {
 	 */
 	public function get_items( $request ) {
+
+		// If $params['active'] is true then only active workflows should be returned.
```
I changed this behavior because otherwise it is only possible to request either active workflows or inactive workflows, but not both together. The UI requires both together. Since the description of the `active` param is "Whether to return only active workflows", I thought this behavior seems to fit.

kallehauge replied:
I made a slight modification to this approach.
This specific code has been reverted to what it was before, but the `active` API endpoint argument has been made optional (read: we've removed the `default` value), so unless anything else is specified, then it will return both active and inactive workflows

robertf4 commented on projects/plugins/crm/src/js/state/automations-admin/util.ts:
```diff
@@ -1,26 +1,15 @@
+ * @param {Workflow} workflow - The workflow to prepare
+ * @returns {ServerPreparedWorkflow} The prepared workflow
  */
-export const findStep = ( workflow: Workflow, stepId: string ) => {
```
I deleted this along with its corresponding tests as it isn't being used.

robertf4 committed:
Change axios header

robertf4 commented on projects/plugins/crm/src/js/components/automations-admin/components/bulk-workflow-actions/index.tsx:
```diff
@@ -1,26 +1,61 @@
+				mutateWorkflows(
+					{ ...workflow, active: true },
+					{
+						onSuccess: () => refetchWorkflows(),
```
This looks like it would refetch workflows after each individual workflow in a bulk action is updated, but it doesn't due to `react-query` magic.

kallehauge replied:
![giphy (32)](https://github.com/Automattic/jetpack/assets/3846700/95044ec9-4b7b-4d0d-9698-a9699365e629)

robertf4 committed:
Make it easier to search for URLs

robertf4 commented on projects/plugins/crm/src/js/components/automations-admin/components/edit-modal/index.tsx:
```diff
@@ -23,15 +30,17 @@ export const EditModal: React.FC< EditModalProps > = ( { isOpen, onClose, workfl
+
 	const onSave = useCallback( () => {
-		// TODO: save workflow here
+		mutateWorkflows( workflow );
```
If I refetch here on success then the state doesn't change. If I refetch on failure then the user's changes are lost with no indication that it happened. Thus this save will fail silently until we implement error notifications.

kallehauge committed:
Make "active" parameter optional

kallehauge committed:
Correct admin page padding on smaller screens

kallehauge committed:
Improve responsiveness of table and overall alignment

cleacos committed:
Handle any Automation exception processing each workflow. Skip, show and continue.

A workflow could have an error, if an exception happens, it stop all the process.
We should let the system process all the workflow and show any errer instead of stop on error.

Signed-off-by: cleacos <javacodia@gmail.com>

kallehauge committed:
Add tooltip to trigger column

cleacos committed:
Add more available fields to the Contact_Field_Changed condition.

Signed-off-by: cleacos <javacodia@gmail.com>

cleacos committed:
The fullname is not a valid field in the Contact Entity.

Signed-off-by: cleacos <javacodia@gmail.com>

kallehauge committed:
Merge branch 'crm/add/automation-saving-and-loading' of github.com:Automattic/jetpack into crm/add/automation-saving-and-loading

kallehauge committed:
Show all triggers in workflow table

Practically speaking we're only supporting one trigger right now, but in theory we could have more which is why we store and send
an array of triggers.

kallehauge committed:
changelog

kallehauge committed:
Show all triggers in edit modal

kallehauge committed:
Add trigger/step labels

This is copied from the "build your own" experience, but it significantly improves immediate understanding of the edit modal as
well.

kallehauge committed:
Merge branch 'crm/update/automations-action-attributes' into crm/add/automation-saving-and-loading
